lib/filter-cloudwatch-compiler.js

```diff
@@ -68,80 +78,187 @@ var FilterCloudWatchCompiler = ASTVisitor.extend({
         var left, right, ret;
 
         switch (node.operator) {
 
-            case 'OR':
+            case 'AND':
                 left = this.visit(node.left);
                 right = this.visit(node.right);
 
-                ret = _.extend({}, right, left);
+                // You can only combine single conditions
+                if (left.length > 1 || right.length > 1) {
+                    this.throwUnsupportedFilter(node.location, 'AND between anything other than simple conditions');
+                }
+                left = left[0];
+                right = right[0];
+
+                // You can only combine items if one of the products/items/metrics
+                // is wildcarded (undefined for products, an empty array for items/metrics).
+                if (left.product !== undefined &&
+                    right.product !== undefined) {
+                    this.throwUnsupportedFilter(node.location, 'AND between products');
+                }
+
+                if ((left.item.length > 0 && right.item.length > 0)) {
+                    this.throwUnsupportedFilter(node.location, 'AND between items');
+                }
+
+                if ((left.metric.length > 0 && right.metric.length > 0)) {
+                    this.throwUnsupportedFilter(node.location, 'AND between metrics');
+                }
+
+                // merge the two halves by combining the objects.
+                ret = [{
+                    product: (left.product === undefined ? right.product : left.product),
+                    metric: _.union(left.metric, right.metric),
+                    item: _.union(left.item, right.item),
+                }];
+
+                break;
+
+            case 'OR':
+                left = this.visit(node.left);
+                right = this.visit(node.right);
 
-                // The extend() didn't actually handle the case where
-                // left and right both had a value for the same key.
-                // Handle that here by merging the two arrays.
-                ret = _.mapObject(ret, function(val, key) {
-                    return _.union(val, right[key]);
-                });
+                // Just concatenate the conditions. We'll try to merge at the end.
+                ret = left.concat(right);
 
-                return ret;
+                break;
 
             case '==':
                 left = this.visit(node.left);
                 right = this.visit(node.right);
 
-                // Left *must* be either 'item' or 'product'. When
-                // left is 'product', right must be one of the
-                // supported AWS products.
-                if (left !== 'item' && left !== 'product') {
+                // Left *must* be either 'item', 'product', or
+                // 'metric'. When left is 'product', right must be one
+                // of the supported AWS products.
+                if (left !== 'item' && left !== 'product' && left !== 'metric') {
```

Comment:
Seems like a ```const CONDITION_KEYS = ['product', 'item', 'metric']``` at the top of the file might be useful.

Then `CONDITION_KEYS.indexOf(left) >= 0`